packages/ozone/src/scheduled-action/service.ts

```diff
@@ -0,0 +1,276 @@
+      .selectFrom('scheduled_action')
+      .selectAll()
+      .where('did', '=', did)
+      .where('action', '=', action)
+      .where('status', '=', 'pending')
+      .executeTakeFirst()
+
+    return scheduledAction || null
+  }
+
+  async listScheduledActions({
+    cursor,
+    limit = 50,
+    startTime,
+    endTime,
+    subjects,
+    statuses,
+    direction = 'desc',
+  }: {
+    cursor?: string
+    limit?: number
+    startTime?: Date
+    endTime?: Date
+    subjects?: string[]
+    statuses?: ScheduledActionStatus[]
+    direction?: 'asc' | 'desc'
+  } = {}): Promise<{
+    actions: Selectable<ScheduledAction>[]
+    cursor?: string
+  }> {
+    let query = this.db.db.selectFrom('scheduled_action').selectAll()
+
+    if (subjects && subjects.length > 0) {
+      query = query.where('did', 'in', subjects)
+    }
+
+    if (statuses && statuses.length > 0) {
+      query = query.where('status', 'in', statuses)
+    }
+
+    if (startTime) {
+      query = query.where((qb) => {
+        return qb
+          .orWhere('executeAt', '>=', startTime.toISOString())
+          .orWhere('executeAfter', '>=', startTime.toISOString())
+      })
+    }
+
+    if (endTime) {
+      query = query.where((qb) => {
+        return qb
+          .orWhere('executeAt', '<=', endTime.toISOString())
+          .orWhere('executeUntil', '<=', endTime.toISOString())
+          .orWhere((sqb) => {
+            return sqb
+              .where('executeUntil', 'is', null)
+              .where('executeAfter', '<=', endTime.toISOString())
+          })
+      })
+    }
+
+    if (cursor) {
+      query = query.where(
+        'id',
+        direction === 'asc' ? '>' : '<',
+        parseInt(cursor, 10),
+      )
+    }
+
+    const actions = await query.orderBy('id', direction).limit(limit).execute()
+
+    return {
+      actions,
+      cursor: actions.at(-1)?.id?.toString(),
+    }
+  }
+
+  async cancelScheduledActions(subjects: string[]): Promise<{
+    succeeded: string[]
+    failed: { did: string; error: string; errorCode?: string }[]
+  }> {
+    const succeeded: string[] = []
+    const failed: { did: string; error: string; errorCode?: string }[] = []
+
+    for (const did of subjects) {
+      try {
+        const result = await this.db.db
+          .updateTable('scheduled_action')
+          .set({
+            status: 'cancelled',
+            updatedAt: new Date().toISOString(),
+          })
+          .where('did', '=', did)
+          .where('status', '=', 'pending')
+          .executeTakeFirst()
+
+        if (result.numUpdatedRows && result.numUpdatedRows > 0) {
+          succeeded.push(did)
+        } else {
+          failed.push({
+            did,
+            error: 'No pending scheduled actions found for subject',
+            errorCode: 'NoPendingActions',
+          })
+        }
+      } catch (error) {
+        failed.push({
+          did,
+          error: error instanceof Error ? error.message : 'Unknown error',
```

Comment:
For unexpected operational errors it's typically best practice not to expose them through the API since they can leak sensitive details.